espnet2/lm/espnet_model_multitask.py

```diff
@@ -0,0 +1,208 @@
+    LabelSmoothingLoss,
+)
+
+
+class ESPnetMultitaskLanguageModel(AbsESPnetModel):
+    def __init__(
+        self,
+        lm: AbsLM,
+        vocab_size: int,
+        token_list: Union[Tuple[str, ...], List[str]],
+        ignore_id: int = 0,
+        lsm_weight: float = 0.0,
+        length_normalized_loss: bool = False,
+        sos_syms: List[str] = ["<generatetext>", "<generatespeech>"],
+        eos_sym: str = "<sos/eos>",
+    ):
+        assert check_argument_types()
+        super().__init__()
+        self.lm = lm
+        self.sos_ids = [token_list.index(t) for t in sos_syms]
+        self.eos_id = token_list.index(eos_sym)
+
+        # ignore_id may be assumed as 0, shared with CTC-blank symbol for ASR.
+        self.ignore_id = ignore_id
+
+        self.token_list = token_list.copy()
+
+        self.criterion_att = LabelSmoothingLoss(
+            size=vocab_size,
+            padding_idx=ignore_id,
+            smoothing=lsm_weight,
+            normalize_length=length_normalized_loss,
+        )
+
+    def nll(
+        self,
+        text: torch.Tensor,
+        text_lengths: torch.Tensor,
+        max_length: Optional[int] = None,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        """Compute negative log likelihood (nll)
+        NOTE(yifan): We only use nll to calculate perplexity,
+            so there is no condition in each sentence.
+
+        Normally, this function is called in batchify_nll.
+        Args:
+            text: (Batch, Length)
+            text_lengths: (Batch,)
+            max_lengths: int
+        """
+        assert max_length is None
+
+        batch_size = text.size(0)
+        # For data parallel
+        if max_length is None:
+            text = text[:, : text_lengths.max()]
+        else:
+            text = text[:, :max_length]
+
+        # NOTE(yifan): The first token is space when using bpe
+        text = text[:, 1:]
+        text_lengths = text_lengths - 1
+
+        # 1. Create a sentence pair like '<sos> w1 w2 w3' and 'w1 w2 w3 <eos>'
+        # text: (Batch, Length) -> x, y: (Batch, Length + 1)
+        x, x_lengths = text, text_lengths  # text already has <sos>
+        t = F.pad(text, [0, 1], "constant", self.ignore_id)
+        for i, l in enumerate(text_lengths):
+            t[i, l] = self.eos_id
+        t = t[:, 1:]  # remove <sos>
+
+        # 2. Forward Language model
+        # x: (Batch, Length) -> y: (Batch, Length, NVocab)
+        y, _ = self.lm(x, None)
+
+        # 3. Calc negative log likelihood
+        # nll: (BxL,)
+        nll = F.cross_entropy(y.view(-1, y.shape[-1]), t.view(-1), reduction="none")
+        # nll: (BxL,) -> (BxL,)
+        if max_length is None:
+            nll.masked_fill_(make_pad_mask(x_lengths).to(nll.device).view(-1), 0.0)
+        else:
+            # nll.masked_fill_(
+            # make_pad_mask(x_lengths, maxlen=max_length + 1).to(nll.device).view(-1),
+            #     0.0,
+            # )
```

Comment:
Do you want to keep them?
If so, please leave some comments of why do you want to keep them.

Reply:
Removed